test/extended/csrapprover/csrapprover.go

```diff
@@ -0,0 +1,163 @@
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		certRequestTemplate := &x509.CertificateRequest{
+			SignatureAlgorithm: x509.ECDSAWithSHA256,
+			Subject: pkix.Name{
+				CommonName:   "system:node:hacking-node.ec2.internal",
+				Organization: []string{"system:nodes"},
+			},
+			PublicKey: priv.PublicKey,
+		}
+
+		csrbytes, err := x509.CreateCertificateRequest(rand.Reader, certRequestTemplate, priv)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		csrSpec := certv1beta1.CertificateSigningRequestSpec{
+			Request: pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrbytes}),
+			Usages: []certv1beta1.KeyUsage{
+				certv1beta1.UsageDigitalSignature,
+				certv1beta1.UsageKeyEncipherment,
+				certv1beta1.UsageClientAuth,
+			},
+			Groups: []string{
+				"system:serviceaccounts",
+				"system:serviceaccounts:openshift-machine-config-operator",
+				"system:authenticated",
+			},
+		}
+
+		kubeCSRString, err := getKubeCSRObjectString("some-csr", csrSpec)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// we simulate having the token by running the 'CSR create' as the node-bootstrapper SA
+		err = oc.AsAdmin().
+			Run("create").
+			Args("--as", "system:serviceaccount:openshift-machine-config-operator:node-bootstrapper", "-f", "-").
+			InputString(kubeCSRString).
+			Execute()
+
+		csrClient := oc.AdminKubeClient().CertificatesV1beta1().CertificateSigningRequests()
+		defer cleanupCSR(csrClient, "some-csr")
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		err = waitCSRStatus(csrClient)
+		o.Expect(err).NotTo(o.HaveOccurred())
+	})
+})
+
+func getKubeCSRObjectString(csrName string, spec certv1beta1.CertificateSigningRequestSpec) (string, error) {
+	s := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme.Scheme, scheme.Scheme)
+	mcsr := &certv1beta1.CertificateSigningRequest{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "CertificateSigningRequest",
+			APIVersion: "certificates.k8s.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: csrName,
+		},
+		Spec: spec,
+	}
+
+	var kubeCSRBytes bytes.Buffer
+	byteWriter := bufio.NewWriter(&kubeCSRBytes)
+	err := s.Encode(mcsr, byteWriter)
+	if err != nil {
+		return "", err
+	}
+
+	err = byteWriter.Flush()
+	return kubeCSRBytes.String(), err
+}
+
+// waits for the CSR object to change status, checks that it did not get approved
+func waitCSRStatus(csrAdminClient certclientv1beta1.CertificateSigningRequestInterface) error {
+	stopCh := make(chan struct{})
+
+	return wait.PollUntil(30*time.Second, func() (bool, error) {
+		csr, err := csrAdminClient.Get("some-csr", metav1.GetOptions{})
+		if !apierrors.IsNotFound(err) && err != nil {
+			return false, err
+		}
+		if len(csr.Status.Conditions) > 0 {
+			for _, c := range csr.Status.Conditions {
+				if c.Type == certv1beta1.CertificateApproved {
+					return true, fmt.Errorf("CSR for unknown node should not be approved")
+				} else if c.Type == certv1beta1.CertificateDenied {
```

Comment:
remove this

Reply:
Right, the machine approver never denies